### PR TITLE
Add util cogs

### PR DESCRIPTION
**Pull request by ChrisLovering**
This PR adds a bunch of utility cogs to the bot, copy and pasted from the python bot with minor modifications/removing unneeded features

### CODE REVIEW

File: bot/exts/utils/error_handler.py [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+    def __init__(self, bot: ThreadBot):
+        self.bot = bot
+
+    def _get_error_embed(self, title: str, body: str) -> Embed:
+        """Return an embed that contains the exception."""
+        return Embed(
+            title=title,
+            colour=Colours.error,
+            description=body
+        )
```
Comment (Bluenix2) [inline bot/exts/utils/error_handler.py]: Should this perhaps be a static method?

File: bot/exts/utils/extensions.py [hunk 2]
```diff
@@ -0,0 +1,259 @@
...
+        """Only allow moderators and core developers to invoke the commands in this cog."""
+        return await commands.has_any_role(*MODERATION_ROLES, Roles.core_developers).predicate(ctx)
+
+    # This cannot be static (must have a __func__ attribute).
+    async def cog_command_error(self, ctx: Context, error: Exception) -> None:
+        """Handle BadArgument errors locally to prevent the help command from showing."""
+        if isinstance(error, commands.BadArgument):
+            await ctx.send(str(error))
+            return
+        logger.error(f"Unhandled error in ext cog: {error}")
```
Comment (Bluenix2) [inline bot/exts/utils/extensions.py]: I'd prefer these at the top of the cog

File: bot/exts/utils/source.py [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+from pathlib import Path
+from typing import Optional, Tuple, Union
+
+from discord import Embed, utils
+from discord.ext import commands
+
+from bot.bot import ThreadBot
+from bot.constants import URLs
+
+SourceType = Union[commands.HelpCommand, commands.Command, commands.Cog, str, commands.ExtensionNotLoaded]
```
Comment (Bluenix2) [inline bot/exts/utils/source.py]: Isn't `commands.ExtensionNotLoaded` an error?

File: bot/exts/utils/source.py [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+        cmd = ctx.bot.get_command(argument)
+        if cmd:
+            return cmd
+
+        tags_cog = ctx.bot.get_cog("Tags")
+        show_tag = True
+
+        if not tags_cog:
+            show_tag = False
+        elif argument.lower() in tags_cog._cache:
```
Comment (Bluenix2) [inline bot/exts/utils/source.py]:
```suggestion
        show_tag = True if tags_cog else False

        if argument.lower() in tags_cog._cache:
```
Which is just..
```suggestion
        show_tag = bool(tags_cog)

        if argument.lower() in tags_cog._cache:
```
Which simply can be removed and just use `if tags_cog`.

File: bot/exts/utils/source.py [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+            return argument.lower()
+
+        escaped_arg = utils.escape_markdown(argument)
+
+        raise commands.BadArgument(
+            f"Unable to convert '{escaped_arg}' to valid command{', tag,' if show_tag else ''} or Cog."
```
Comment (Bluenix2) [inline bot/exts/utils/source.py]:
```suggestion
            f"Unable to convert '{escaped_arg}' to valid command{', tag,' if tags_cog else ''} or Cog."
```

**Review by Bluenix2 — COMMENTED** [review]
Mostly skimmed through it, looks good

File: bot/exts/utils/error_handler.py (hunk 1 above)
Comment (ChrisLovering) [reply bot/exts/utils/error_handler.py]: 2a5960d

File: bot/exts/utils/source.py (hunk 4 above)
Comment (ChrisLovering) [reply bot/exts/utils/source.py]: 46bb863

File: bot/exts/utils/source.py (hunk 5 above)
Comment (ChrisLovering) [reply bot/exts/utils/source.py]: 46bb863

File: bot/exts/utils/source.py (hunk 3 above)
Comment (ChrisLovering) [reply bot/exts/utils/source.py]: This was some code I missed when removing unneded things when copying from the bot. ext not loaded is reaised in that command then you try to get the source of a tag and the tags cog isn't loaded.

3003ac9

**Review by Bluenix2 — APPROVED** [review]
Again, quickly skimmed through it. I see nothing wrong, though I haven't tested it.

**Pull request merged**